packages/widget-v2/src/components/RenderWalletList.tsx

```diff
@@ -54,71 +53,132 @@ export const RenderWalletListHeader = ({
-      );
+  const theme = useTheme();
+  const modal = useModal();
+  const connectMutation = useMutation({
+    mutationKey: ["connectWallet"],
+    mutationFn: async (wallet: MinimalWallet) => {
+      return await wallet.connect();
     },
-    [onSelect]
-  );
+    onSuccess: () => {
+      modal.remove();
+    }
+  });
+
+  const renderItem = (wallet: Wallet) => {
+    const {
+      walletName,
+      walletPrettyName,
+      walletInfo: { logo: imageUrl },
+      rightContent,
+    } = wallet;
+    return (
+      <ModalRowItem
+        key={walletName}
+        onClick={() => {
+          connectMutation.mutate(wallet);
+        }}
+        style={{ marginTop: ITEM_GAP }}
+        leftContent={
+          <Row align="center" gap={10}>
+            {imageUrl && (
+              <img
+                height={35}
+                width={35}
+                style={{ objectFit: "cover" }}
+                src={imageUrl}
+                alt={`${walletPrettyName} logo`}
+              />
+            )}
+
+            <Text>{walletPrettyName}</Text>
+          </Row>
+        }
+        rightContent={rightContent?.()}
+      />
+    );
+  };
 
   const height = useMemo(() => {
     return Math.min(530, walletList.length * (ITEM_HEIGHT + ITEM_GAP));
   }, [walletList]);
 
+  const container = useMemo(() => {
+    if (connectMutation.isError) {
+      return (
+        <StyledInnerContainer height={height} >
+          <StyledLoadingContainer>
+            <StyledAnimatedBorder
+              width={80}
+              height={80}
+              backgroundColor={theme.primary.text.normal}
+              txState="failed"
+              borderSize={8}
+            >
+              <img
+                style={{ objectFit: "cover" }}
+                src={connectMutation.variables?.walletInfo.logo}
+                alt={`${connectMutation.variables?.walletPrettyName} logo`} />
+            </StyledAnimatedBorder>
+            <Text color={theme.primary.text.lowContrast}>Failed to connect to {connectMutation.variables?.walletPrettyName}</Text>
+          </StyledLoadingContainer>
+        </StyledInnerContainer >
+      );
+    }
+
+    if (connectMutation.isPending) {
+      return (
+        <StyledInnerContainer height={height}>
+          <StyledLoadingContainer>
+            <StyledAnimatedBorder
+              width={80}
+              height={80}
+              backgroundColor={theme.primary.text.normal}
+              txState="broadcasted"
+              borderSize={8}
+            >
+              <img
+                style={{ objectFit: "cover" }}
+                src={connectMutation.variables?.walletInfo.logo}
+                alt={`${connectMutation.variables?.walletPrettyName} logo`} />
+            </StyledAnimatedBorder>
+            <Text color={theme.primary.text.lowContrast}>Connecting to {connectMutation.variables?.walletPrettyName}</Text>
+          </StyledLoadingContainer>
+        </StyledInnerContainer>
+      );
+    }
```

Comment:
@codingki these two are very similar, is there a state that is true for both of them? I think we can combine them and then just use different txState and text for the "connecting to" and "failed to connect to", since everything else is exactly the same

Reply:
sure